### ISSUE TITLE
Pass `rustc` the equivalent of `-Wframe-larger-than=1000`

### ISSUE THREAD

**Issue by ojeda** [Rust-for-Linux/linux#121]

**Comment by mpe** [comment]
Is there such a thing? I had a look and couldn't find it, but I'm not a rustc expert.

**Comment by ojeda** [comment]
Not sure -- we may need to implement it ourselves! :)

**Comment by bjorn3** [comment]
No, there isn't a direct option, but there are things that can help. There is https://github.com/rust-lang/rust/pull/83519 for moves of large values. This includes assignments. For stack frame sizes, only LLVM knows about it. With `-Z emit-stack-sizes` it is possible to tell LLVM to add a section containing info about stack frame sizes. This option was added to rustc for https://github.com/japaric/cargo-call-stack.

**Comment by mpe** [comment]
OK. I used `-Z emit-stack-sizes` to debug issue #155, which worked well enough to find the large functions. I guess that's not useful for an upstream stack size warning option though, because it would require nightly rustc (AIUI).

**Comment by TheSven73** [comment]
@mpe would you mind sharing how you obtained a list of function stack sizes using `-Z emit-stack-sizes`? I'd like to do the same for arm 32-bit. It's quite possible that stack overflows are happening there too.

**Comment by mpe** [comment]
Something like this:
```
$ make KRUSTCFLAGS="-Z emit-stack-sizes"
$ cargo install stack-sizes
$ stack-sizes vmlinux | sort -nk 2
```

**Comment by ojeda** [comment]
> I guess that's not useful for an upstream stack size warning option though, because it would require nightly rustc (AIUI).

Yeah -- we should use it nevertheless. That way the `rustc` folks know it is useful and being used (see the #2 list), and that we will want it stabilized at some point. If really needed, we can always gate it or take it out later.

**Comment by bjorn3** [comment]
I don't know what stability guarantee LLVM has for this feature. If LLVM doesn't guarantee that it will remain the same forever, I think it is unlikely that it will be stabilized in rustc.

**Comment by ojeda** [comment]
My thinking was that given Clang supports `-Wframe-larger-than=`, LLVM will need to support the feature one way or another, which means it should be fairly easy for `rustc` to keep providing it too.

**Comment by bjorn3** [comment]
Clang doesn't use the stack size section. Instead it instructs LLVM to emit a diagnostic whenever the stack size is above a certain threshold. Rustc doesn't have any support for handling LLVM diagnostics except for inline asm.

https://github.com/llvm/llvm-project/blob/a9968c0a339ab47f5d6da1c513b2754f1e98a596/llvm/lib/CodeGen/PrologEpilogInserter.cpp#L278-L284

https://github.com/llvm/llvm-project/commit/728c5540eef2622fa30c0ca42727b111ff182ce1

**Comment by nickdesaulniers** [comment]
cc @arndb
